tests/functional/backend/RemoteControlTest.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace ls\tests;
+
+use Yii;
+
+class RemoteControlTest extends TestBaseClassWeb
+{
+    private static $tmpBaseUrl;
+    private static $tmpRPCType;
+
+    private static $client;
+
+    public static function setUpBeforeClass(): void
+    {
+        parent::setUpBeforeClass();
+
+        $urlMan = Yii::app()->urlManager;
+        self::$tmpBaseUrl = $urlMan->getBaseUrl();
+        $urlMan->setBaseUrl('http://' . self::$domain . '/index.php');
+        //$serverUrl = App()->createAbsoluteUrl('/admin/remotecontrol');
+        $serverUrl = $urlMan->createUrl('/admin/remotecontrol');
+
+        self::$tmpRPCType = Yii::app()->getConfig('RPCInterface');
+
+        if (self::$tmpRPCType === 'off') {
+            \SettingGlobal::setSetting('RPCInterface', 'json');
+            $RPCType = 'json';
+        } else {
+            $RPCType = self::$tmpRPCType;
+        }
+
+        if ($RPCType == 'xml') {
+            $cur_path = get_include_path();
+            set_include_path($cur_path . PATH_SEPARATOR . APPPATH . 'helpers');
+            require_once('Zend/XmlRpc/Client.php');
+
+            self::$client = new \Zend_XmlRpc_Client($serverUrl);
+        } elseif ($RPCType == 'json') {
+            Yii::app()->loadLibrary('jsonRPCClient');
+            self::$client = new \jsonRPCClient($serverUrl);
+        }
+    }
+
+    public static function tearDownAfterClass(): void
+    {
+        parent::tearDownAfterClass();
+
+        $urlMan = Yii::app()->urlManager;
+        $urlMan->setBaseUrl(self::$tmpBaseUrl);
+
+        \SettingGlobal::setSetting('RPCInterface', self::$tmpRPCType);
+    }
+
+    public function testGetSessionKey()
+    {
+        $sessionKey = self::$client->call('get_session_key', ['admin', 'password']);
+        $this->assertIsNotArray($sessionKey);
```

Comment:
[assertIsString()](https://docs.phpunit.de/en/9.6/assertions.html#assertisstring) ?

A 500 error is not an array , right ?

Reply:
I believe a 500 would be shown as an exception. Not sure.
Should try maybe...

Reply:
Very edge case. Does it worth spending on it?

Reply:
> Very edge case. Does it worth spending on it?

The good result **MUST** be a string : then just check if it's a string.

Reply:
This is what `get_session_key` returns.
```
        if (is_string($loginResult)) {
            return array('status' => $loginResult);
        }
        return array('status' => 'Invalid user name or password');
```

Reply:
Also, jsonRpcClient, if something fails, throws Exception.
I would just leave it like this.

Reply:
> This is what `get_session_key` returns.
> 
> ```
>         if (is_string($loginResult)) {
>             return array('status' => $loginResult);
>         }
>         return array('status' => 'Invalid user name or password');
> ```

In case of error only 

In case of success: 
https://github.com/LimeSurvey/LimeSurvey/blob/5d97f49a06b610b32db927d434de76e88e1746a1/application/helpers/remotecontrol/remotecontrol_handle.php#L60


Else : if there are a CRSF issue : it's an Exception.

Reply:
You are right